finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala

```diff
@@ -0,0 +1,91 @@
+package com.twitter.finagle.netty4.transport.buoyant
+
+import com.twitter.concurrent.AsyncQueue
+import com.twitter.finagle.ChannelException
+import com.twitter.finagle.netty4.transport.ChannelTransport
+import com.twitter.util.{Future, Promise}
+import io.netty.channel.ChannelFuture
+import io.netty.{channel => nettyChan}
+import java.util
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * A Transport implementation based on Netty's Channel which buffers writes.  This Transport buffers
+ * writes to a queue and schedules an event on the channel's event loop to write the items and then
+ * flush them.  This allows writes to be batched and flushed together rather than needing to flush
+ * after each individual write.  The result is fewer syscalls which leads to better performance.
+ * This approach was highly inspired by grpc-java's WriteQueue:
+ * https://github.com/grpc/grpc-java/pull/431
+ *
+ * @param ch the underlying netty channel
+ *
+ * @param readQueue the queue used to buffer inbound messages
+ *
+ * @note During the construction, a `ChannelTransport` inserts the terminating
+ *       inbound channel handler into the channel's pipeline so any inbound channel
+ *       handlers inserted after that won't get any of the inbound traffic.
+ */
+class BufferingChannelTransport(
+  ch: nettyChan.Channel,
+  readQueue: AsyncQueue[Any] = new AsyncQueue[Any],
+  omitStackTraceOnInactive: Boolean = false
+) extends ChannelTransport(ch, readQueue, omitStackTraceOnInactive) {
+
+  // Satisfy the done promise when the write completes.
+  private case class WriteItem(msg: Any, done: Promise[Unit])
+
+  // Always flush after this many messages.
+  private[this] val MaxFlushSize = 128
+  private[this] val flushScheduled = new AtomicBoolean(false)
+  private[this] val writeQueue = new LinkedBlockingQueue[WriteItem]()
+  private[this] val writeChunk = new util.ArrayDeque[WriteItem](MaxFlushSize)
+
+  override def write(msg: Any): Future[Unit] = {
+    val p = new Promise[Unit]()
+    writeQueue.add(WriteItem(msg, p))
+    scheduleFlush()
```

Comment:
Does `scheduleFlush()` need to happen when `writeQueue` has reached `MaxFlushSize` of 128? Does this need to be guarded against a condition?

Reply:
I don't think so.  All writes will add the msg to the writeQueue and schedule a flush if one has not already been scheduled.  During the scheduled flush, we will drain off the writeQueue 128 items at a time and write them out as a group.